internal/httputil/httpapi.go

```diff
@@ -89,6 +115,56 @@ func MakeAuthAPI(
 	return MakeExternalAPI(metricsName, h)
 }
 
+func checkConsent(ctx context.Context, userID string, userAPI userapi.UserInternalAPI, userConsentCfg config.UserConsentOptions) error {
+	localPart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return nil
+	}
+	// check which version of the policy the user accepted
+	res := &userapi.QueryPolicyVersionResponse{}
+	err = userAPI.QueryPolicyVersion(ctx, &userapi.QueryPolicyVersionRequest{
+		LocalPart: localPart,
+	}, res)
+	if err != nil {
+		return nil
+	}
+
+	// user hasn't accepted any policy, block access.
+	if userConsentCfg.Version != res.PolicyVersion {
```

Comment:
How does this work for rollbacks? E.g 1.2 != 1.0 but they have consented to a newer version?

Reply:
Not sure how we'd want to fix this, as the version can be a random string. We could use [semver](https://pkg.go.dev/golang.org/x/mod/semver), but that would require prefixing the version with a leading "v"

> Package semver implements comparison of semantic version strings. In this package, semantic version strings must begin with a leading "v", as in "v1.0.0".